blocksub/blocksub.go

```diff
@@ -0,0 +1,196 @@
+// Package blocksub implements an Ethereum block subscriber that works with either a websocket or a polling or both.
+package blocksub
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/log"
+	"go.uber.org/atomic"
+)
+
+type BlockSub struct {
+	PollTimeout time.Duration // 10 seconds by default (8,640 requests per day)
+	SubTimeout  time.Duration // 40 seconds by default, after this timeout the subscriber will reconnect
+
+	ethNodeHTTPURI      string // usually port 8545
+	ethNodeWebsocketURI string // usually port 9546
+
+	headerC chan<- *ethtypes.Header
+	ctx     context.Context
```

Comment:
Go Prefer contexts passed as arguments. https://go.dev/blog/context-and-structs

Reply:
in this case the context could be just passed into `Start`->`runListener` which makes sense.

Reply:
Would be even better to allow multiple subscribers, each with it's own context

Reply:
There's two ways of handling contexts now: 
1. one struct-wide context to close the various loops (main listener, poller, websocket client subscription)
2. individual contexts per subscription